rfcs/45-stageless.md

```diff
@@ -0,0 +1,564 @@
+you can safely parallelize the rest of the work into seperate systems, improving both performance and maintainability.
+
+### States
+
+**States** are one particularly common and powerful form of run criteria, allowing you to toggle systems on-and-off based on the value of a given resource and smoothly handle transitions between states by running cleanup and initialization systems.
+Typically, these are used for relatively complex, far-reaching facets of the game like pausing, loading assets or entering into menus.
+
+The current value of each state is stored in a resource, which must derive the `State` trait.
+These are typically (but not necessarily) enums, where each distinct state is represented as an enum variant.
+
+Each state is associated with three sets of systems:
+
+1. **Update systems:** these systems run each frame if and only if the value of the state resource matches the provided value.
+   1. `app.add_system(my_system.in_state(GameState::Playing))`
+2. **On-enter systems:** these systems run once when the specified state is entered.
+   1. `app.add_system(my_system.on_enter(GameState::Playing))`
+3. **On-exit systems:** these systems run once when the specified stated is exited.
+   1. `app.add_system(my_system.on_enter(GameState::Playing))`
+
+Update systems are by far the simplest: they're simply powered by run criteria.
+`.in_state` is precisely equivalent to `run_if_resource_is`, except with an additional trait bound that the resource must implement `State`.
+
+On-enter and on-exit systems are stored in dedicated schedules, two per state, within the `App's` `Schedules`.
+These schedules can be configured in all of the ordinary ways, but, as they live in different schedules, ordering cannot be defined relative to systems in the main schedule.
+
+Due to their disruptive and far-reaching effects, state transitions do not occur immediately.
+Instead, they are deferred (like commands), until the next `flush_state<S: State>` exclusive system runs.
+This system first runs the `on_exit` schedule of the previous state on the world, then runs the `on_enter` schedule of the new state on the world.
+Once that is complete, the exclusive system ends and control flow resumes as normal.
+
+When states are added using `App::add_state::<S: State>(initial_state)`, one `flush_state<S>` system is added to the app, with the `GeneratedLabel::StateTransition<S>` label.
+You can configure when and if this system is scheduled by configuring this label, and you can add additional copies of this system to your schedule where you see fit.
+Just like with commands, **state-flushed ordering constraints** can be used to verify that state transitions have run at the appropriate time.
+If system `A` is `before_and_flush_state::<S>` system `B`, the schedule will be unsatisfiable unless there is an intervening `flush_state<S>` system.
+
+Apps can have multiple orthogonal states representing independent facets of your game: these operate fully independently.

```

Comment:
> Apps can have multiple orthogonal states representing independent facets of your game

I really like this feature, I thought about using it a while ago, and I actually don't know if it's possible to do such a thing right now.

Reply:
It is possible right now, but not clearly documented :) States with nested enum variants don't work though :(